hazelcast-sql/src/main/java/com/hazelcast/sql/impl/expression/FieldAccessExpression.java

```diff
@@ -76,33 +78,29 @@ public T eval(Row row, ExpressionEvalContext context) {
     @Override
     public T eval(final Row row, final ExpressionEvalContext context, boolean useLazyDeserialization) {
         // Use lazy deserialization for nested queries. Only the last access should be eager.
-        final Object res = ref.eval(row, context, true);
-        if (res == null) {
+        final Object result = ref.eval(row, context, true);
+        if (result == null) {
             return null;
         }
 
-        if (isPrimitive(res.getClass())) {
+        if (isPrimitive(result.getClass())) {
             throw QueryException.error("Field Access expression can not be applied to primitive types");
         }
 
-        if (getterCache.get() == null) {
-            getterCache.compareAndSet(null, new EvictableGetterCache(
-                    MAX_CLASS_COUNT,
-                    MAX_GETTER_PER_CLASS_COUNT,
-                    GetterCache.EVICTABLE_CACHE_EVICTION_PERCENTAGE,
-                    false
-            ));
-        }
-
-        // defensive check, should never happen
-        final GetterCache cache = getterCache.get();
-        assert cache != null : "GetterCache should never be null";
+        getterCache.compareAndSet(null, new EvictableGetterCache(
+                MAX_CLASS_COUNT,
+                MAX_GETTER_PER_CLASS_COUNT,
+                GetterCache.EVICTABLE_CACHE_EVICTION_PERCENTAGE,
+                false
+        ));
 
-        final Extractors extractors = Extractors.newBuilder(context.getSerializationService())
-                .setGetterCacheSupplier(() -> cache)
-                .build();
         try {
-            return (T) type.convert(extractors.extract(res, name, useLazyDeserialization));
+            Object value = result instanceof GenericRecord
+                    ? AvroQueryTarget.extractValue((GenericRecord) result, name)
```

Comment:
is hardcoding `AvroQueryTarget` here a hack/workaround? Is it supposed to stay this way?

Reply:
The field extraction logic for Avro is not part of `Extractors` and it will probably never be since `Getter`s are only used by Predicate API, which is deprecated. The functionality is part of `AvroQueryTarget` and I think we don't need additional encapsulation to make it look less hacky.

Reply:
if we make Avro serializable using SerializationService (next PR), then nothing is going to stop users from putting avro records in imaps and running PredicateAPI queries on them. We might consider supporting that for consistency.
Also SQL uses predicate API queries for nested loops joins, so if IMap MAPPING with avro serialization format could be created, it might need such support.

Reply:
https://hazelcast.atlassian.net/browse/HZ-3161